File: backend/src/main/kotlin/top/bearingwall/asya/controller/UserController.kt
package top.bearingwall.asya.controller

import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import io.jsonwebtoken.JwtException
import jakarta.servlet.http.HttpServletRequest
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.web.PageableDefault
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseCookie
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.CookieValue
import org.springframework.web.bind.annotation.*
import top.bearingwall.asya.dto.BizCode
import top.bearingwall.asya.dto.Result
import top.bearingwall.asya.dto.UserRegistrationRequest
import top.bearingwall.asya.dto.UserResponse
import top.bearingwall.asya.dto.UserInfoResponse
import top.bearingwall.asya.dto.UserUpdateRequest
import top.bearingwall.asya.dto.BatchRegisterRequest
import top.bearingwall.asya.dto.BatchRegisterResponse
import top.bearingwall.asya.dto.TokenRefreshResponse
import top.bearingwall.asya.model.UserRole
import top.bearingwall.asya.service.SystemConfigService
import top.bearingwall.asya.service.UserService
import java.util.UUID
import java.time.Duration

@RestController
@RequestMapping("/api/users")
@Tag(name = "用户管理", description = "用户注册等相关接口")
class UserController(
    private val userService: UserService,
    private val systemConfigService: SystemConfigService
) {
    companion object {
        private const val REFRESH_COOKIE_NAME = "asya_refresh_token"
        private val REFRESH_COOKIE_MAX_AGE: Duration = Duration.ofDays(30)
    }

    @Operation(
        summary = "用户注册",
        description = "创建新用户。密码会使用 BCrypt 进行加密后存储。"
    )
    @PostMapping("/register")
    fun register(
        requestContext: HttpServletRequest,
        @RequestBody request: UserRegistrationRequest
    ): ResponseEntity<Result<UserResponse>> {
        return try {
            val response = userService.registerUser(request)
            ResponseEntity.status(HttpStatus.CREATED)
                .header(HttpHeaders.AUTHORIZATION, "Bearer ${response.response.token}")
                .header(HttpHeaders.SET_COOKIE, buildRefreshCookie(response.refreshToken, requestContext).toString())
                .body(Result.success(response.response))
        } catch (e: IllegalArgumentException) {
            ResponseEntity.status(HttpStatus.OK)
                .body(Result.failure(BizCode.USER_EXISTS, e.message ?: BizCode.USER_EXISTS.message))
        } catch (e: IllegalStateException) {
            ResponseEntity.status(HttpStatus.OK)
                .body(Result.failure(BizCode.PARAM_ERROR, e.message ?: BizCode.PARAM_ERROR.message))
        } catch (e: Exception) {
            handleException(e)
        }
    }

    @Operation(
        summary = "用户登录",
        description = "使用用户名和密码登录，成功后返回用户信息与 Bearer Token。"
    )
    @PostMapping("/login")
    fun login(
        requestContext: HttpServletRequest,
        @RequestBody request: UserRegistrationRequest
    ): ResponseEntity<Result<UserResponse>> {
        return try {
            val response = userService.loginUser(request)
            ResponseEntity.status(HttpStatus.OK)
                .header(HttpHeaders.AUTHORIZATION, "Bearer ${response.response.token}")
                .header(HttpHeaders.SET_COOKIE, buildRefreshCookie(response.refreshToken, requestContext).toString())
                .body(Result.success(response.response))
        } catch (_: IllegalStateException) {
            ResponseEntity.status(HttpStatus.OK)
                .body(Result.failure(BizCode.USER_NOT_FOUND, "系统中还没有该用户，请先完成注册"))
        } catch (e: IllegalArgumentException) {
            ResponseEntity.status(HttpStatus.OK)
                .body(Result.failure(BizCode.PASSWORD_ERROR, e.message ?: BizCode.PASSWORD_ERROR.message))
        } catch (e: Exception) {
            handleException(e)
        }
    }

    @Operation(summary = "刷新 access token", description = "使用 HttpOnly refresh token cookie 换取新的 access token，并轮换 refresh token")
    @PostMapping("/refresh")
    fun refresh(
        requestContext: HttpServletRequest,
        @CookieValue(name = REFRESH_COOKIE_NAME, required = false) refreshToken: String?
    ): ResponseEntity<Result<TokenRefreshResponse>> {
        return try {
            if (refreshToken.isNullOrBlank()) {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Result.failure(BizCode.TOKEN_INVALID, "Refresh token 不存在"))
            }

            val response = userService.refreshAccessToken(refreshToken)
            ResponseEntity.ok()
                .header(HttpHeaders.AUTHORIZATION, "Bearer ${response.response.token}")
                .header(HttpHeaders.SET_COOKIE, buildRefreshCookie(response.refreshToken, requestContext).toString())
                .body(Result.success(response.response))
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .header(HttpHeaders.SET_COOKIE, clearRefreshCookie(requestContext).toString())
                .body(Result.failure(BizCode.TOKEN_INVALID, e.message ?: "Refresh token 无效"))
        }
    }

    @Operation(summary = "退出登录", description = "清除 refresh token cookie")
    @PostMapping("/logout")
    fun logout(
        requestContext: HttpServletRequest
    ): ResponseEntity<Result<Unit>> {
        return ResponseEntity.ok()
            .header(HttpHeaders.SET_COOKIE, clearRefreshCookie(requestContext).toString())
            .body(Result.success(Unit))
    }

    @Operation(summary = "批量注册用户", description = "仅 SYS_ADMIN 可访问，批量注册代表并关联到会议")
    @PostMapping("/batch")
    fun batchRegister(
        @RequestHeader(HttpHeaders.AUTHORIZATION) authorization: String,
        @RequestBody request: BatchRegisterRequest
    ): ResponseEntity<Result<BatchRegisterResponse>> {
        return try {
            val requester = userService.getUserFromToken(extractBearer(authorization))
            val requesterUuid = requester.uuid ?: throw IllegalStateException("Requester id missing")
            val response = userService.batchRegister(requesterUuid, request)
            ResponseEntity.ok(Result.success(response))
        } catch (e: SecurityException) {
            ResponseEntity.status(HttpStatus.FORBIDDEN).body(Result.failure(BizCode.TOKEN_INVALID, e.message ?: "Access Denied"))
        } catch (e: IllegalArgumentException) {
            ResponseEntity.status(HttpStatus.OK).body(Result.failure(BizCode.PARAM_ERROR, e.message ?: "参数错误"))
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.OK).body(Result.failure(BizCode.TOKEN_INVALID, e.message ?: "Error"))
        }
    }

    @Operation(summary = "分页查询用户", description = "仅系统管理员可访问，可按昵称、显示名称、关联会议、角色筛选")
    @GetMapping
    fun listAll(
        @RequestHeader(HttpHeaders.AUTHORIZATION) authorization: String,
        @RequestParam(required = false) name: String?,
        @RequestParam(required = false) displayName: String?,
        @RequestParam(required = false) conferenceUuid: UUID?,
        @RequestParam(required = false) role: UserRole?,
        @RequestParam(required = false) current: Int?,
        @RequestParam(required = false) pageNum: Int?,
        @PageableDefault(sort = ["name"], direction = Sort.Direction.ASC) pageable: Pageable
    ): ResponseEntity<Result<Page<UserInfoResponse>>> {
        return try {
            val user = userService.getUserFromToken(extractBearer(authorization))
            if (user.role != UserRole.SYS_ADMIN) {
                return ResponseEntity.status(HttpStatus.OK)
                    .body(Result.failure(BizCode.TOKEN_INVALID, "需要管理员权限"))
            }

            val effectivePageable = resolvePageable(pageable, current, pageNum)
            val users = userService.getUsers(
                pageable = effectivePageable,
                name = name,
                displayName = displayName,
                conferenceUuid = conferenceUuid,
                role = role
            )
            ResponseEntity.ok(Result.success(users))
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.OK)
                .body(Result.failure(BizCode.TOKEN_INVALID, e.message ?: "Token解析失败"))
        }
    }

    @Operation(summary = "获取当前登录用户信息")
    @GetMapping("/user")
    fun getCurrentUser(
        @RequestHeader(HttpHeaders.AUTHORIZATION) authorization: String
    ): ResponseEntity<Result<UserInfoResponse>> {
        return try {
            val token = extractBearer(authorization)
            val info = userService.getCurrentUserInfo(token)
            ResponseEntity.ok(Result.success(info))
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.OK)
                .body(Result.failure(BizCode.TOKEN_INVALID, e.message ?: "Token解析失败"))
        }
    }

    @Operation(summary = "更新用户信息", description = "管理员可修改所有用户；普通用户仅能修改自己的信息")
    @PutMapping("/user/{uuid}")
    fun updateUser(
        @PathVariable uuid: UUID,
        @RequestHeader(HttpHeaders.AUTHORIZATION) authorization: String,
        @RequestBody request: UserUpdateRequest
    ): ResponseEntity<Result<UserInfoResponse>> {
        return try {
            val token = extractBearer(authorization)
            val updated = userService.updateUser(uuid, token, request)
            ResponseEntity.ok(Result.success(updated))
        } catch (e: Exception) {
            if (e is JwtException) {
                return ResponseEntity.status(HttpStatus.OK)
                    .body(Result.failure(BizCode.TOKEN_INVALID, e.message ?: "Token失效"))
            }
            val status = if (e is IllegalArgumentException && e.message == "Permission denied") {
                HttpStatus.FORBIDDEN
            } else {
                HttpStatus.OK
            }
            ResponseEntity.status(status)
                .body(Result.failure(BizCode.PARAM_ERROR, e.message ?: "Update failed"))
        }
    }

    @Operation(summary = "管理员重置用户密码", description = "仅系统管理员可执行")
    @PostMapping("/{uuid}/password-reset")
    fun resetPassword(
        @PathVariable uuid: UUID,
        @RequestHeader(HttpHeaders.AUTHORIZATION) authorization: String,
        @RequestBody body: Map<String, String>
    ): ResponseEntity<Result<Unit>> {
        return try {
            val requester = userService.getUserFromToken(extractBearer(authorization))
            if (requester.role != UserRole.SYS_ADMIN) {
                return ResponseEntity.status(HttpStatus.FORBIDDEN)
                    .body(Result.failure(BizCode.PERMISSION_DENIED, "需要管理员权限"))
            }

            val newPassword = body["password"]
            if (newPassword.isNullOrBlank()) {
                throw IllegalArgumentException("密码不能为空")
            }

            userService.resetPassword(uuid, newPassword)

            ResponseEntity.ok(Result.success(Unit))
        } catch (e: Exception) {
            handleException(e)
        }
    }

    @Operation(summary = "删除用户", description = "仅系统管理员可执行")
    @DeleteMapping("/{uuid}")
    fun deleteUser(
        @PathVariable uuid: UUID,
        @RequestHeader(HttpHeaders.AUTHORIZATION) authorization: String
    ): ResponseEntity<Result<Unit>> {
        return try {
            val requester = userService.getUserFromToken(extractBearer(authorization))
            if (requester.role != UserRole.SYS_ADMIN) {
                return ResponseEntity.status(HttpStatus.FORBIDDEN)
                    .body(Result.failure(BizCode.PERMISSION_DENIED, "需要管理员权限"))
            }

            userService.deleteUser(uuid)
            ResponseEntity.ok(Result.success(Unit))
        } catch (e: Exception) {
            handleException(e)
        }
    }

    @Operation(summary = "设置是否允许注册", description = "仅系统管理员可执行。true: 允许, false: 禁止")
    @PostMapping("/registration-switch")
    fun setRegistrationSwitch(
        @RequestHeader(HttpHeaders.AUTHORIZATION) authorization: String,
        @RequestParam allowed: Boolean
    ): ResponseEntity<Result<Boolean>> {
        return try {
            val requester = userService.getUserFromToken(extractBearer(authorization))
            if (requester.role != UserRole.SYS_ADMIN) {
                return ResponseEntity.status(HttpStatus.FORBIDDEN)
                    .body(Result.failure(BizCode.PERMISSION_DENIED, "需要管理员权限"))
            }

            systemConfigService.setRegistrationAllowed(allowed)
            ResponseEntity.ok(Result.success(allowed))
        } catch (e: Exception) {
            handleException(e)
        }
    }

    @Operation(summary = "查询当前是否允许注册")
    @GetMapping("/registration-switch")
    fun getRegistrationSwitch(): ResponseEntity<Result<Boolean>> {
        return try {
            val allowed = userService.isRegistrationAvailable()
            ResponseEntity.ok(Result.success(allowed))
        } catch (e: Exception) {
            handleException(e)
        }
    }

    private fun extractBearer(authorization: String): String {
        val prefix = "Bearer "
        if (!authorization.startsWith(prefix)) {
            throw IllegalArgumentException("Authorization header must start with 'Bearer '")
        }
        return authorization.substring(prefix.length)
    }

    private fun resolvePageable(pageable: Pageable, current: Int?, pageNum: Int?): Pageable {
        val oneBasedPage = current ?: pageNum ?: return pageable
        val zeroBasedPage = (oneBasedPage - 1).coerceAtLeast(0)
        return PageRequest.of(zeroBasedPage, pageable.pageSize, pageable.sort)
    }

    private fun buildRefreshCookie(refreshToken: String, request: HttpServletRequest): ResponseCookie {
        return ResponseCookie.from(REFRESH_COOKIE_NAME, refreshToken)
            .httpOnly(true)
            .secure(isSecureRequest(request))
            .sameSite("Lax")
            .path("/api/users")
            .maxAge(REFRESH_COOKIE_MAX_AGE)
            .build()
    }

    private fun clearRefreshCookie(request: HttpServletRequest): ResponseCookie {
        return ResponseCookie.from(REFRESH_COOKIE_NAME, "")
            .httpOnly(true)
            .secure(isSecureRequest(request))
            .sameSite("Lax")
            .path("/api/users")
            .maxAge(Duration.ZERO)
            .build()
    }

    private fun isSecureRequest(request: HttpServletRequest): Boolean {
        return request.isSecure || request.getHeader("X-Forwarded-Proto")?.equals("https", ignoreCase = true) == true
    }

    private fun <T> handleException(e: Exception): ResponseEntity<Result<T>> {
        if (e is JwtException) {
            return ResponseEntity.status(HttpStatus.OK)
                .body(Result.failure(BizCode.TOKEN_INVALID, e.message ?: "Token失效"))
        }
        return ResponseEntity.status(HttpStatus.OK)
            .body(Result.failure(BizCode.PARAM_ERROR, e.message ?: "操作失败"))
    }
}

File: backend-dotnet/src/AsyaMun.Api/Controllers/UserController.cs
using AsyaMun.Api.Auth;
using AsyaMun.Api.Audit;
using AsyaMun.Api.Dtos;
using AsyaMun.Api.Errors;
using AsyaMun.Api.Models;
using AsyaMun.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace AsyaMun.Api.Controllers;

[ApiController]
[Route("api/users")]
public class UserController : ControllerBase
{
    private readonly UserService _userService;
    private readonly SystemConfigService _systemConfigService;

    public UserController(UserService userService, SystemConfigService systemConfigService)
    {
        _userService = userService;
        _systemConfigService = systemConfigService;
    }

    [HttpPost("register")]
    [AllowAnonymous]
    [Auditable(AuditActionType.USER_REGISTER, "用户注册")]
    public async Task<IActionResult> Register([FromBody] UserRegistrationRequest? request)
    {
        if (request == null)
        {
            return ApiResp.Fail(BizCode.PARAM_ERROR, "请求体不能为空");
        }

        var response = await _userService.RegisterUserAsync(request);
        Response.WriteBearer(response.Response.Token);
        RefreshCookieHelper.Append(HttpContext, response.RefreshToken);

        return ApiResp.Ok(response.Response, StatusCodes.Status201Created);
    }

    [HttpPost("login")]
    [AllowAnonymous]
    [Auditable(AuditActionType.USER_LOGIN, "用户登录")]
    public async Task<IActionResult> Login([FromBody] UserRegistrationRequest? request)
    {
        if (request == null)
        {
            return ApiResp.Fail(BizCode.PARAM_ERROR, "请求体不能为空");
        }

        var response = await _userService.LoginUserAsync(request);
        Response.WriteBearer(response.Response.Token);
        RefreshCookieHelper.Append(HttpContext, response.RefreshToken);

        return ApiResp.Ok(response.Response);
    }

    [HttpPost("refresh")]
    [AllowAnonymous]
    public async Task<IActionResult> Refresh()
    {
        var refreshToken = RefreshCookieHelper.Read(HttpContext);
        if (string.IsNullOrWhiteSpace(refreshToken))
        {
            RefreshCookieHelper.Clear(HttpContext);
            return ApiResp.Fail(BizCode.TOKEN_INVALID, "Refresh token 不存在", StatusCodes.Status401Unauthorized);
        }

        try
        {
            var response = await _userService.RefreshAccessTokenAsync(refreshToken);
            Response.WriteBearer(response.Response.Token);
            RefreshCookieHelper.Append(HttpContext, response.RefreshToken);

            return ApiResp.Ok(response.Response);
        }
        catch (Exception ex)
        {
            RefreshCookieHelper.Clear(HttpContext);
            return ApiResp.Fail(BizCode.TOKEN_INVALID, ex.Message ?? "Refresh token 无效", StatusCodes.Status401Unauthorized);
        }
    }

    [HttpPost("logout")]
    [AllowAnonymous]
    public IActionResult Logout()
    {
        RefreshCookieHelper.Clear(HttpContext);
        return ApiResp.OkUnit();
    }

    [HttpPost("batch")]
    [Authorize]
    [Auditable(AuditActionType.USER_BATCH_REGISTER, "批量注册用户")]
    public async Task<IActionResult> Batch([FromBody] BatchRegisterRequest? request)
    {
        if (request == null)
        {
            return ApiResp.Fail(BizCode.PARAM_ERROR, "请求体不能为空");
        }

        var requester = RequireCurrentUser();
        var response = await _userService.BatchRegisterAsync(requester, request);

        return ApiResp.Ok(response);
    }

    [HttpPost("batch-full")]
    [Authorize]
    [Auditable(AuditActionType.USER_BATCH_REGISTER_FULL, "批量注册用户（含角色与会组）")]
    public async Task<IActionResult> BatchFull([FromBody] BatchRegisterFullRequest? request)
    {
        if (request == null)
        {
            return ApiResp.Fail(BizCode.PARAM_ERROR, "请求体不能为空");
        }

        var requester = RequireCurrentUser();
        var response = await _userService.BatchRegisterFullAsync(requester, request);

        return ApiResp.Ok(response);
    }

    [HttpGet]
    [Authorize]
    public async Task<IActionResult> ListAll(
        [FromQuery] string? name,
        [FromQuery] string? displayName,
        [FromQuery] Guid? conferenceUuid,
        [FromQuery] UserRole? role)
    {
        var requester = RequireCurrentUser();
        if (requester.Role != UserRole.SYS_ADMIN)
        {
            return ApiResp.Fail(BizCode.TOKEN_INVALID, "需要管理员权限");
        }

        var input = PageInput.Parse(HttpContext.Request, 20);
        var page = await _userService.GetUsersAsync(input, name, displayName, conferenceUuid, role);

        return ApiResp.Ok(page);
    }

    [HttpGet("user")]
    [Authorize]
    public IActionResult GetCurrentUser()
    {
        var requester = RequireCurrentUser();
        return ApiResp.Ok(_userService.GetCurrentUserInfo(requester));
    }

    [HttpPut("user/{uuid:guid}")]
    [Authorize]
    [Auditable(AuditActionType.USER_UPDATE, "更新用户信息")]
    public async Task<IActionResult> UpdateUser(Guid uuid, [FromBody] UserUpdateRequest? request)
    {
        if (request == null)
        {
            return ApiResp.Fail(BizCode.PARAM_ERROR, "请求体不能为空");
        }

        var requester = RequireCurrentUser();
        var updated = await _userService.UpdateUserAsync(uuid, requester, request);

        return ApiResp.Ok(updated);
    }

    [HttpPost("{uuid:guid}/password-reset")]
    [Authorize]
    [Auditable(AuditActionType.USER_PASSWORD_RESET, "重置用户密码")]
    public async Task<IActionResult> ResetPassword(Guid uuid, [FromBody] ResetPasswordBody? body)
    {
        var requester = RequireCurrentUser();
        if (requester.Role is not (UserRole.SYS_ADMIN or UserRole.DH))
        {
            return ApiResp.Fail(BizCode.PERMISSION_DENIED, "需要管理员或主席团指导权限", StatusCodes.Status403Forbidden);
        }

        var newPassword = body?.Password;
        if (string.IsNullOrWhiteSpace(newPassword))
        {
            return ApiResp.Fail(BizCode.PARAM_ERROR, "密码不能为空");
        }

        await _userService.ResetPasswordAsync(uuid, newPassword, requester.Uuid);

        return ApiResp.OkUnit();
    }

    [HttpDelete("{uuid:guid}")]
    [Authorize]
    [Auditable(AuditActionType.USER_DELETE, "删除用户")]
    public async Task<IActionResult> DeleteUser(Guid uuid)
    {
        var requester = RequireCurrentUser();
        if (requester.Role != UserRole.SYS_ADMIN)
        {
            return ApiResp.Fail(BizCode.PERMISSION_DENIED, "需要管理员权限", StatusCodes.Status403Forbidden);
        }

        await _userService.DeleteUserAsync(uuid);

        return ApiResp.OkUnit();
    }

    [HttpPost("registration-switch")]
    [Authorize]
    [Auditable(AuditActionType.USER_REGISTRATION_SWITCH, "设置用户注册开关")]
    public async Task<IActionResult> SetRegistrationSwitch([FromQuery] bool allowed)
    {
        var requester = RequireCurrentUser();
        if (requester.Role != UserRole.SYS_ADMIN)
        {
            return ApiResp.Fail(BizCode.PERMISSION_DENIED, "需要管理员权限", StatusCodes.Status403Forbidden);
        }

        await _systemConfigService.SetRegistrationAllowedAsync(allowed);

        return ApiResp.Ok(allowed);
    }

    [HttpGet("registration-switch")]
    [AllowAnonymous]
    public async Task<IActionResult> GetRegistrationSwitch()
    {
        return ApiResp.Ok(await _userService.IsRegistrationAvailableAsync());
    }

    private User RequireCurrentUser()
    {
        return HttpContext.GetCurrentUser()
            ?? throw new TokenInvalidException("登录状态已失效，请重新登录");
    }
}

public record ResetPasswordBody(string? Password);